tket/src/WeightSubgrMono/Reducing/HallSetReducer.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/Reducing/HallSetReducer.hpp"
+
+#include <algorithm>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+#include "WeightSubgrMono/Searching/SearchBranch.hpp"
+#include "WeightSubgrMono/Searching/SearchNodeWrapper.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+// We put smaller domains at the TOP (back) of a vector, so we can pop_back
+// the smaller domains.
+bool HallSetReducer::VariableData::operator<(const VariableData& other) const {
+  return domain_size > other.domain_size ||
+         (domain_size == other.domain_size && vertex < other.vertex);
+}
+
+bool HallSetReducer::reduce(
+    SearchNodeWrapper& search_node_wrapper, SearchBranch& branch) const {
+  m_domain_sizes_and_vertices.clear();
+
+  // Initial fill of the data.
+  for (const auto& entry :
+       search_node_wrapper.get().pattern_v_to_possible_target_v) {
+    m_domain_sizes_and_vertices.emplace_back();
+    m_domain_sizes_and_vertices.back().vertex = entry.first;
+    m_domain_sizes_and_vertices.back().domain_size = entry.second.size();
+
+    if (m_domain_sizes_and_vertices.back().domain_size == 0) {
+      return false;
+    }
+  }
+  std::sort(
+      m_domain_sizes_and_vertices.begin(), m_domain_sizes_and_vertices.end());
+
+  for (;;) {
+    const auto old_size = m_domain_sizes_and_vertices.size();
+    if (!find_and_remove_top_hall_set_block(search_node_wrapper, branch)) {
+      return false;
+    }
+    if (old_size == m_domain_sizes_and_vertices.size()) {
+      break;
+    }
+    TKET_ASSERT(old_size > m_domain_sizes_and_vertices.size());
```

Comment:
If in doubt, assert more! Obviously every assert is unnecessary, if code is without errors!

(Any "necessary" assert would itself BE an error!)

Reply:
Of course, asserts should be cheap. (E.g., there's an example where we'd LIKE to assert the X is a subset of Y, but this would be expensive. Hence, we merely assert we X.size() <= Y.size(), and check that ONE element of X is contained within Y).

Also, it's amazing in the early stages how many apparently simple asserts quickly detect nasty bugs, and how often it's completely unexpected.